### ISSUE TITLE
Bunny::Channel#close should throw immediately if the channel is already closed by the server

### ISSUE THREAD

**Issue by carlhoerberg** [ruby-amqp/bunny#528]
```ruby
require 'bunny'

b = Bunny.new.start
ch1 = b.create_channel
ch1.queue("q1").subscribe(exclusive: true) {|d,h,p| }
ch2 = b.create_channel

begin
  ch2.queue("q1").subscribe(exclusive: true) {|d,h,p| }
rescue => e
  puts e.inspect
  ch2.close
end
```

Result:
```
#<Bunny::AccessRefused: ACCESS_REFUSED - queue 'q1' in vhost '/' in exclusive use>
W, [2017-11-06T15:49:06.381848 #62510]  WARN -- #<Bunny::Session:0x7fd2f2863748 guest@127.0.0.1:5672, vhost=/, addresses=[127.0.0.1:5672]>: Recovering from connection.close (CHANNEL_ERROR - expected 'channel.open')
W, [2017-11-06T15:49:06.382057 #62510]  WARN -- #<Bunny::Session:0x7fd2f2863748 guest@127.0.0.1:5672, vhost=/, addresses=[127.0.0.1:5672]>: Will recover from a network failure (no retry limit)...
/Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/concurrent/continuation_queue.rb:39:in `block in poll': Timeout::Error (Timeout::Error)
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/concurrent/continuation_queue.rb:30:in `synchronize'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/concurrent/continuation_queue.rb:30:in `poll'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/session.rb:1343:in `wait_on_continuations'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/session.rb:537:in `block in close_channel'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/session.rb:533:in `close_channel'
        from /Users/carl/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/bunny-2.7.1/lib/bunny/channel.rb:236:in `close'
        from bug.rb:12:in `rescue in <main>'
        from bug.rb:8:in `<main>'

shell returned 1
```

**Comment by michaelklishin** [comment]
This is a race condition we cannot avoid entirely. Server-sent `channel.close` does not arrive immediately.

**Comment by carlhoerberg** [comment]
I don't know, even adding `sleep 1` between rescue and `ch2.close` raises the exception?

**Comment by michaelklishin** [comment]
[Bunny::Channel#close](https://github.com/ruby-amqp/bunny/blob/master/lib/bunny/channel.rb#L235) doesn't throw early the way most other clients do.

**Comment by michaelklishin** [comment]
@carlhoerberg I'm curious to know if the solution I ended up with in https://github.com/ruby-amqp/bunny/commit/9df7cb04d9ff12b1af62a11e239fd81e5472c872 makes sense to you. If so I'm happy to release a 2.8.0 very soon.

Thanks again for reporting the issues you encounter.

**Comment by carlhoerberg** [comment]
Super nice!

**Comment by michaelklishin** [comment]
Shipped in `2.8.0`.
